app/src/main/kotlin/com/github/gotify/messages/MessagesViewModel.kt

```diff
@@ -0,0 +1,210 @@
+package com.github.gotify.messages
+
+import androidx.lifecycle.ViewModel
+import androidx.lifecycle.viewModelScope
+import com.github.gotify.ApplicationState
+import com.github.gotify.Repository
+import com.github.gotify.client.model.Application
+import com.github.gotify.client.model.Message
+import com.github.gotify.messages.provider.MessageWithImage
+import java.util.concurrent.atomic.AtomicBoolean
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.SharingStarted
+import kotlinx.coroutines.flow.StateFlow
+import kotlinx.coroutines.flow.combine
+import kotlinx.coroutines.flow.distinctUntilChanged
+import kotlinx.coroutines.flow.flatMapLatest
+import kotlinx.coroutines.flow.map
+import kotlinx.coroutines.flow.onEach
+import kotlinx.coroutines.flow.stateIn
+import kotlinx.coroutines.launch
+
+/**
+ * ViewModel for the messages screen.
+ *
+ * @param repository The repository to get the data from and to send the data to.
+ */
+internal class MessagesViewModel(private val repository: Repository) : ViewModel() {
+    private val loadingMore = AtomicBoolean(false)
+    private val shadowDeletedMessages: MutableStateFlow<Set<Long>> = MutableStateFlow(emptySet())
+    private val currentApp = MutableStateFlow<Application?>(null)
+    private val _refreshing = MutableStateFlow(false)
+
+    /**
+     * Whether the messages are currently being refreshed.
+     */
+    val refreshing = _refreshing
```

Comment:
this should probably remove mutability. 
```suggestion
    val refreshing: StateFlow<Boolean> = _refreshing
```